synapse/handlers/sync.py

```diff
@@ -980,7 +980,18 @@ def _get_rooms_changed(self, sync_result_builder, ignored_users):
             # We want to figure out if we joined the room at some point since
             # the last sync (even if we have since left). This is to make sure
             # we do send down the room, and with full state, where necessary
+
             old_state_ids = None
+            if room_id in joined_room_ids and non_joins:
+                # Always include if the user (re)joined the room, especially
+                # important so that device list changes are calculated correctly.
+                # If there are non join member events, but we are still in the room,
+                # then the user must have joined (or left and joined)
```

Comment:
surely this *only* happens if the user has left and joined

or conceivably been invited and joined